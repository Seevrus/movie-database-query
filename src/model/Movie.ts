import { MovieResult } from '../types/search-movies';

export class Movie {
  id?: number;

  title?: string;

  originalTitle?: string;

  posterPath?: string | null;

  isAdult?: boolean;

  genreIds?: number[];

  overview?: string;

  popularity?: number;

  voteAverage?: number;

  voteCount?: number;

  releaseDate?: string;

  originalLanguage?: string;

  video?: boolean;

  backdropPath?: string;

  constructor(movieData: MovieResult) {
    this.id = movieData.id;
    this.title = movieData.title;
    this.originalTitle = movieData.original_title;
    this.posterPath = movieData.poster_path;
    this.isAdult = movieData.adult;
    this.genreIds = movieData.genre_ids;
    this.overview = movieData.overview;
    this.popularity = movieData.popularity;
    this.voteAverage = movieData.vote_average;
    this.voteCount = movieData.vote_count;
    this.releaseDate = movieData.release_date;
    this.originalLanguage = movieData.original_language;
    this.video = movieData.video;
    this.backdropPath = movieData.backdrop_path;
  }
}
