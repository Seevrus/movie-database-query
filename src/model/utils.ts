export function jsonFetch<ResponseT>(
  url: string,
  config: RequestInit = {},
): Promise<ResponseT> {
  return fetch(url, config)
    .then((response) => response.json())
    .then((data) => data as ResponseT);
}
