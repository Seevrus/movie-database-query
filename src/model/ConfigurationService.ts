import { ApiConfigurationResponseT, ApiConfigurationT } from '../types/api-configuration';
import { jsonFetch } from './utils';

export const defaultConfiguration = {
  baseUrl: 'https://image.tmdb.org/t/p/',
  size: 'w185',
};

export class ConfigurationService {
  static async getConfiguration(): Promise<ApiConfigurationT> {
    return jsonFetch<ApiConfigurationResponseT>('https://api.themoviedb.org/3/configuration', {
      headers: {
        Authorization: `Bearer ${process.env.REACT_APP_API_TOKEN}`,
      },
    })
      .then(response => ({
        baseUrl: response.images.secure_base_url ?? response.images.base_url,
        size: 'w185',
      }))
      .catch(() => defaultConfiguration);
  }
}