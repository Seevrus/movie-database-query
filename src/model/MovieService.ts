import { SearchMoviesT } from '../types/search-movies';
import { Movie } from './Movie';
import { jsonFetch } from './utils';

export class MovieService {
  static async searchMovies(query: string) {
    return jsonFetch<SearchMoviesT>(`https://api.themoviedb.org/3/search/movie?language=en-US&query=${query}&page=1&include_adult=false`, {
      headers: {
        Authorization: `Bearer ${process.env.REACT_APP_API_TOKEN}`,
      },
    }).then((response) => {
      return (response.results ?? []).map(movieResult => new Movie(movieResult));
    });
  }
}
