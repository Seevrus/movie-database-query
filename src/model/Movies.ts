import { SearchMoviesT } from '../types/search-movies';
import { Movie } from './Movie';
import { jsonFetch } from './utils';

export class Movies {
  movies: Movie[];

  constructor() {
    this.movies = [];
  }

  async searchMovies(query: string) {
    return jsonFetch<SearchMoviesT>(`https://api.themoviedb.org/3/search/movie?language=en-US&query=${query}&page=1&include_adult=false`, {
      headers: {
        Authorization: `Bearer ${process.env.REACT_APP_API_TOKEN}`,
      },
    }).then((response) => {
      this.movies = (response.results ?? []).map(movieResult => new Movie(movieResult));
    });
  }
}
