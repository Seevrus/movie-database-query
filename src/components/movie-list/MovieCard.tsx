import {
  MDBCard,
  MDBCardTitle,
  MDBCardText,
  MDBCardBody,
  MDBCardImage,
  MDBRow,
  MDBCol,
} from 'mdb-react-ui-kit';
import { useContext } from 'react';
import { Button } from 'react-bootstrap';
import { MovieListContext } from './MovieListContext';
import { Movie } from '../../model/Movie';

import './movie-card.css';

interface MovieCardProps {
  movie: Movie;
}

export const MovieCard = ({ movie }: MovieCardProps) => {
  const { configuration, favoriteMovies, setFavoriteMovies } = useContext(MovieListContext);
  const hasImage = !!configuration.size && !!movie.posterPath;
  const imgUrl = hasImage
    ? configuration.baseUrl + '/' + configuration.size + '/' + movie.posterPath
    : '';

  const addMovieToFavorites = (m: Movie) => {
    setFavoriteMovies?.([
      ...favoriteMovies,
      m,
    ]);
  };

  const isMovieFavorite = (m: Movie) => {
    return favoriteMovies.findIndex((m2) => m.id === m2.id) !== -1;
  };

  const removeFromFavorites = (m: Movie) => {
    setFavoriteMovies?.(favoriteMovies.filter((m2) => m.id !== m2.id));
  };

  return (
    <MDBCard style={{ maxWidth: '1280px' }}>
      <MDBRow className='g-0'>
        <MDBCol md='3'>
          <MDBCardImage src={imgUrl} alt={`Poster of ${movie.title}`} fluid />
        </MDBCol>
        <MDBCol md='9'>
          <MDBCardBody>
            <MDBCardTitle>{movie.title}</MDBCardTitle>
            <MDBCardText>
              {movie.overview}
            </MDBCardText>
            <MDBCardText>
              <small className='text-muted'>{movie.releaseDate}</small>
            </MDBCardText>
            {isMovieFavorite(movie)
              ? <Button
                  onClick={() => removeFromFavorites(movie)}
                >Eltávolítás a kedvencek közül</Button>
              : <Button
                  variant="light"
                  onClick={() => addMovieToFavorites(movie)}
                >Kedvencek közé</Button>
            }
          </MDBCardBody>
        </MDBCol>
      </MDBRow>
    </MDBCard>
  );
};
