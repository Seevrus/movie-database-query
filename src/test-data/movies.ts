import { Movie } from '../model/Movie';

const movies = [
  new Movie({
    id: 1,
    title: 'Reality Bites',
  }),
  new Movie({
    id: 2,
    title: 'Demolition Man',
  }),
  new Movie({
    id: 3,
    title: 'Kramer vs. Kramer',
  }),
  new Movie({
    id: 4,
    title: 'The Christmas Chronicles: Part Two',
  }),
  new Movie({
    id: 5,
    title: 'The Ides of March',
  }),
  new Movie({
    id: 6,
    title: 'The Princess Switch 3: Romancing the Star',
  }),
];

export default movies;
