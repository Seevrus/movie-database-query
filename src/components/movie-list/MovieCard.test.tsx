import renderer from 'react-test-renderer';
import React from 'react';
import { MovieCard } from './MovieCard';
import { Movie } from '../../model/Movie';
import movies from '../../test-data/movies';

describe('MovieCard', () => {
  let contextSpy: jest.SpyInstance;
  const movie = new Movie({
    id: 1984,
    overview: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur ultricies molestie massa, sit amet mattis libero aliquet in. Proin id nunc nec sapien feugiat aliquam a sed erat. Suspendisse cursus nulla et felis varius rutrum. Aliquam tincidunt risus eget rhoncus dapibus. Donec a augue pellentesque, ultrices dolor quis, pulvinar ligula. Nunc dictum, dui in congue molestie, mi justo laoreet tortor, eget interdum lectus urna et mauris. In fringilla quam sem, ut posuere nisl luctus id.',
    poster_path: 'path/to/poster',
    release_date: '2027-07-07',
    title: 'Swimming Pool',
  });

  beforeEach(() => {
    contextSpy = jest.spyOn(React, 'useContext').mockReturnValue({
      configuration: {
        baseUrl: 'http://abc.de',
        size: 'w007',
      },
      favoriteMovies: movies,
      setFavoriteMovies: jest.fn(),
    });
  });

  afterAll(() => {
    contextSpy.mockRestore();
  });

  it('should render correctly', () => {
    const component = renderer.create(
      <MovieCard movie={movie} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render correctly if the movie is a favorite', () => {
    movie.id = 1;

    const component = renderer.create(
      <MovieCard movie={movie} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
