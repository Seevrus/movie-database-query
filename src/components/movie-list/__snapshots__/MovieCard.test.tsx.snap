// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should render correctly 1`] = `
<div
  className="card"
  style={
    Object {
      "maxWidth": "1280px",
    }
  }
>
  <div
    className="row g-0"
  >
    <div
      className="col-md-3"
    >
      <img
        alt="Poster of Swimming Pool"
        className="img-fluid"
        src="http://abc.de/w007/path/to/poster"
      />
    </div>
    <div
      className="col-md-9"
    >
      <div
        className="card-body"
      >
        <h5
          className="card-title"
        >
          Swimming Pool
        </h5>
        <p
          className="card-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur ultricies molestie massa, sit amet mattis libero aliquet in. Proin id nunc nec sapien feugiat aliquam a sed erat. Suspendisse cursus nulla et felis varius rutrum. Aliquam tincidunt risus eget rhoncus dapibus. Donec a augue pellentesque, ultrices dolor quis, pulvinar ligula. Nunc dictum, dui in congue molestie, mi justo laoreet tortor, eget interdum lectus urna et mauris. In fringilla quam sem, ut posuere nisl luctus id.
        </p>
        <p
          className="card-text"
        >
          <small
            className="text-muted"
          >
            2027-07-07
          </small>
        </p>
        <button
          className="btn btn-light"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Kedvencek közé
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`MovieCard should render correctly if the movie is a favorite 1`] = `
<div
  className="card"
  style={
    Object {
      "maxWidth": "1280px",
    }
  }
>
  <div
    className="row g-0"
  >
    <div
      className="col-md-3"
    >
      <img
        alt="Poster of Swimming Pool"
        className="img-fluid"
        src="http://abc.de/w007/path/to/poster"
      />
    </div>
    <div
      className="col-md-9"
    >
      <div
        className="card-body"
      >
        <h5
          className="card-title"
        >
          Swimming Pool
        </h5>
        <p
          className="card-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur ultricies molestie massa, sit amet mattis libero aliquet in. Proin id nunc nec sapien feugiat aliquam a sed erat. Suspendisse cursus nulla et felis varius rutrum. Aliquam tincidunt risus eget rhoncus dapibus. Donec a augue pellentesque, ultrices dolor quis, pulvinar ligula. Nunc dictum, dui in congue molestie, mi justo laoreet tortor, eget interdum lectus urna et mauris. In fringilla quam sem, ut posuere nisl luctus id.
        </p>
        <p
          className="card-text"
        >
          <small
            className="text-muted"
          >
            2027-07-07
          </small>
        </p>
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Eltávolítás a kedvencek közül
        </button>
      </div>
    </div>
  </div>
</div>
`;
